crates/bevy_ecs/src/schedule/graph/graph_map.rs

```diff
@@ -0,0 +1,480 @@
+    where
+        S: Default,
+    {
+        Self::default()
+    }
+
+    /// Create a new `Graph` with estimated capacity.
+    pub(crate) fn with_capacity(nodes: usize, edges: usize) -> Self
+    where
+        S: Default,
+    {
+        Self {
+            nodes: IndexMap::with_capacity_and_hasher(nodes, S::default()),
+            edges: HashSet::with_capacity_and_hasher(edges, S::default()),
+        }
+    }
+
+    /// Use their natural order to map the node pair (a, b) to a canonical edge id.
+    #[inline]
+    fn edge_key(a: NodeId, b: NodeId) -> CompactNodeIdPair {
+        let (a, b) = if DIRECTED || a <= b { (a, b) } else { (b, a) };
+
+        CompactNodeIdPair::store(a, b)
+    }
+
+    /// Return the number of nodes in the graph.
+    pub fn node_count(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// Add node `n` to the graph.
+    pub(crate) fn add_node(&mut self, n: NodeId) {
+        self.nodes.entry(n).or_default();
+    }
+
+    /// Remove a node `n` from the graph.
+    ///
+    /// Computes in **O(V)** time, due to the removal of edges with other nodes.
+    pub(crate) fn remove_node(&mut self, n: NodeId) {
+        let Some(links) = self.nodes.swap_remove(&n) else {
+            return;
+        };
+
+        let links = links.into_iter().map(CompactNodeIdAndDirection::load);
+
+        for (succ, dir) in links {
+            let edge = if dir == Outgoing {
+                Self::edge_key(n, succ)
+            } else {
+                Self::edge_key(succ, n)
+            };
+            // remove all successor links
+            self.remove_single_edge(&succ, &n, dir.opposite());
+            // Remove all edge values
+            self.edges.remove(&edge);
+        }
+    }
+
+    /// Return `true` if the node is contained in the graph.
+    pub fn contains_node(&self, n: NodeId) -> bool {
+        self.nodes.contains_key(&n)
+    }
+
+    /// Add an edge connecting `a` and `b` to the graph.
+    /// For a directed graph, the edge is directed from `a` to `b`.
+    ///
+    /// Inserts nodes `a` and/or `b` if they aren't already part of the graph.
+    pub(crate) fn add_edge(&mut self, a: NodeId, b: NodeId) {
+        if self.edges.insert(Self::edge_key(a, b)) {
+            // insert in the adjacency list if it's a new edge
+            self.nodes
+                .entry(a)
+                .or_insert_with(|| Vec::with_capacity(1))
+                .push(CompactNodeIdAndDirection::store(b, Outgoing));
+            if a != b {
+                // self loops don't have the Incoming entry
+                self.nodes
+                    .entry(b)
+                    .or_insert_with(|| Vec::with_capacity(1))
+                    .push(CompactNodeIdAndDirection::store(a, Incoming));
+            }
+        }
+    }
+
+    /// Remove edge relation from a to b
+    ///
+    /// Return `true` if it did exist.
+    fn remove_single_edge(&mut self, a: &NodeId, b: &NodeId, dir: Direction) -> bool {
```

Comment:
```suggestion
    fn remove_single_edge(&mut self, a: NodeId, b: NodeId, dir: Direction) -> bool {
```
`NodeId` is copy, I think this is better?